FreeRTOS_ARP.c

```diff
@@ -725,17 +725,16 @@ void FreeRTOS_OutputARPRequest( uint32_t ulIPAddress )
 
         #if defined( ipconfigETHERNET_MINIMUM_PACKET_BYTES )
             {
-                if( pxNetworkBuffer->xDataLength < ( size_t ) ipconfigETHERNET_MINIMUM_PACKET_BYTES )
-                {
-                    BaseType_t xIndex;
-
-                    for( xIndex = ( BaseType_t ) pxNetworkBuffer->xDataLength; xIndex < ( BaseType_t ) ipconfigETHERNET_MINIMUM_PACKET_BYTES; xIndex++ )
-                    {
-                        pxNetworkBuffer->pucEthernetBuffer[ xIndex ] = 0U;
-                    }
+                /* At this stage, the xDataLength is would be equal to sizeof( ARPPacket_t )
+                 * due to the call to vARPGenerateRequestPacket above. */
+                BaseType_t xIndex;
 
-                    pxNetworkBuffer->xDataLength = ( size_t ) ipconfigETHERNET_MINIMUM_PACKET_BYTES;
+                for( xIndex = ( BaseType_t ) pxNetworkBuffer->xDataLength; xIndex < ( BaseType_t ) ipconfigETHERNET_MINIMUM_PACKET_BYTES; xIndex++ )
```

Comment:
It is true that `xDataLength >= sizeof( ARPPacket_t )`, but it has not been proven that `xDataLength < ipconfigETHERNET_MINIMUM_PACKET_BYTES`.

An ARP packet is 42 bytes long, the minimum packet length is typically 60 bytes. So 18 bytes must be cleared.

The check for `xDataLength < ipconfigETHERNET_MINIMUM_PACKET_BYTES` is important, otherwise we might truncate longer packets.

**Background**: the macro `ipconfigETHERNET_MINIMUM_PACKET_BYTES` normally has a value of 60 bytes. That is the minimum length of an Ethernet packet. Shorter packets can be sent, but the collision detection might not work correctly. I have also seen that shorter packets got dropped.

Some users commented that the minimum should be 64 bytes: they are right, but the 4 Frame Check Sequence bytes (FCS) are generated automatically by the peripheral. The stack doesn't see them.

Reply:
How would the code reach a point where `xDataLength > ipconfigETHERNET_MINIMUM_PACKET_BYTES`?
The call to vARPGenerateRequestPacket unconditionally modifies the `xDataLength = sizeof( ARPPacket_t )`.
Thus, the value of `xDataLength` would always be less than `ipconfigETHERNET_MINIMUM_PACKET_BYTES`.

Am I missing something here?

Reply:
You are right that `ipconfigETHERNET_MINIMUM_PACKET_BYTES` ought to be longer than `sizeof( ARPPacket_t )`, but it is not necessarily true. If it is defined as 40, we should not trim the packet here:
~~~c
pxNetworkBuffer->xDataLength = ( size_t ) ipconfigETHERNET_MINIMUM_PACKET_BYTES;
~~~

Reply:
By the way, the packet length is being checked in 4 locations: ARP, TCP, UDP, and in the function `vReturnEthernetFrame()` ( FreeRTOS_IP.c ).
It is not much code, but we could create a function that is called from those 4 locations. The logic will remain the same, symbolically:
~~~
if( xDataLength < ipconfigETHERNET_MINIMUM_PACKET_BYTES )
{
    /* Clear the extra bytes. */
    /* Set the new, extended length. */
    xDataLength = ipconfigETHERNET_MINIMUM_PACKET_BYTES;
}
~~~
And only if `ipconfigETHERNET_MINIMUM_PACKET_BYTES` is defined as a macro.